lkl: Add FreeBSD platform support

cemeyer opened the pull request:
Please don't pull this as-is!  In particular, the first few commits are big hacks that worked around issues only present on FreeBSD.

For example, without "Hack NR_CPUS" I got some warning about an invalid range; "timeconst.bc" hacks around limitations in FreeBSD's `bc(1)` program by hardcoding 250 hz in the bc script; and "Build 64-bit" hardcodes CONFIG_64BIT (without which it seemed to try and build 32-bit on a 64-bit host).

Despite those kludges, there are some decent portability-improving patches in here that won't hurt Linux or NT support.  And some fixes for compiler or other issues (e.g. statfs64 was missing a 3rd argument which resulted in `-EINVAL` on all calls).

With this patchset I have ext4 mounting r/w on FreeBSD w/ `lklfuse`.  Neat!

tavip commented:
Awesome ! I will test and cherry pick most of the patches and rework some - "Build 64bit" can use the OUTPUT_FORMAT to detect that host is 64bit and enable 64bit build. Thanks !

cemeyer commented:
Thanks!

tavip commented:
Hi Conrad,

I have created a new branch (lkl_fbsd) which has your fixes. I've added your Signed-off-by: to your commits to adhere with the Linux process, hope that is ok.

I've also squashed a few commits and reworked some. Could you take a look and give it a try? Hope I didn't broke anything :)

Thanks,
Tavi

cemeyer commented:
> I've added your Signed-off-by: to your commits to adhere with the Linux process, hope that is ok.

Fine with me.

> I've also squashed a few commits and reworked some. Could you take a look and give it a try? Hope I didn't broke anything :)

I'll take a look, thanks.

cemeyer commented:
A test build on `lkl_fbsd` branch:

```
...
  CHK     include/generated/timeconst.h
dc: not a string
  UPD     include/generated/timeconst.h
...
In file included from include/linux/jiffies.h:10:0,
                 from include/linux/ktime.h:25,
                 from include/linux/rcupdate.h:47,
                 from include/linux/srcu.h:33,
                 from include/linux/notifier.h:15,
                 from include/linux/memory_hotplug.h:6,
                 from include/linux/mmzone.h:812,
                 from include/linux/gfp.h:5,
                 from include/linux/kmod.h:22,
                 from include/linux/module.h:13,
                 from init/main.c:15:
include/generated/timeconst.h:11:2: error: #error "include/generated/timeconst.h has the wrong HZ value!"
 #error "include/generated/timeconst.h has the wrong HZ value!"
  ^
include/generated/timeconst.h:14:2: error: #error Totally bogus HZ value!
 #error Totally bogus HZ value!
  ^
include/generated/timeconst.h:15:1: error: expected identifier or '(' before numeric constant
 0
 ^
...
```

Yup, the bc script is still a problem.

cemeyer commented:
With just the BC hack from before and a second patch, `lkl_fbsd` builds again: https://github.com/cemeyer/linux/tree/lkl_fbsd_plus

tavip commented:
>Fine with me. (Edit: if it's not too difficult, can the Signed-off-by address be changed to >cem@FreeBSD.org?)

Yep, I'll do that. I guess I should also change the Author address, right?

>Re: "error: ignoring return value of ‘write’, declared with attribute warn_unused_result [-Werror=unused->result]" (both of them) — I believe you can just cast the result to void.

With the new compilers (at least on Ubuntu 14.04) the (void) cast trick does not work anymore apparently due to better optimizations :)

> -largp for fs2tar got dropped from cemeyer@8c2495a),

Oops, I'll fix that.

> I suspect the bc script will still fail on FreeBSD, unfortunately

Not sure how to deal with this. Can we keep it separate for now until maybe we find a better solution? 

I'll rebase the patches to address the above and force push to the same branch and way you say ok I will merge it to the main branch.

cemeyer commented:
> Yep, I'll do that. I guess I should also change the Author address, right?

Sure.

> With the new compilers (at least on Ubuntu 14.04) the (void) cast trick does not work anymore apparently due to better optimizations :)

I think it is just GCC being very pedantic about `warn_unused_result`.

> > I suspect the bc script will still fail on FreeBSD, unfortunately

> Not sure how to deal with this. Can we keep it separate for now until maybe we find a better solution?

Yes.  I think I'll try and import GNU bc into FreeBSD ports.  After that, we may want a patch that changes `bc` invocations to `$(BC)` and sets `BC` to `bc` on Linux and `gbc` on FreeBSD.  I think that's a straightforward way of getting where we want.

> I'll rebase the patches to address the above and force push to the same branch and way you say ok I will merge it to the main branch.

Great, thanks!

tavip commented:
Merged, thanks Conrad !
